libn/i/setmatrix: make generic and constructorless

corhere opened the pull request:
Related to https://github.com/moby/moby/pull/45163#discussion_r1137207454

Allow SetMatrix to be used as a value type with a ready-to-use zero value. SetMatrix values are already non-copyable by virtue of having a mutex field so there is no harm in allowing non-pointer values to be used as local variables or struct fields. Any attempts to pass around by-value copies, e.g. as function arguments, will be flagged by go vet.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

akerouanton commented on libnetwork/network.go:
```diff
@@ -1411,24 +1411,20 @@ func (n *network) addSvcRecords(eID, name, serviceID string, epIP, epIPv6 net.IP
 
 	sr, ok := c.svcRecords[networkID]
```
Unrelated to your PR; do you know why DNS entries are stored in the controller instead of the network? I'm wondering if things could be made the other way around, as the current implementation requires the controller to be locked each and every time a DNS entry is updated or a DNS query is handled. This looks wrong from my point of view.

corhere replied:
It also looks wrong to me. The (predecessor to) `svcRecords` was moved from the network to the controller in https://github.com/moby/libnetwork/commit/25f496f771e66f333ec8153471091fd7f31ffb3e so presumably it had something to do with libnetwork's persistence model and the (now obsolete) reliance on a distributed KV store for propagating state to nodes.

corhere commented on libnetwork/internal/setmatrix/setmatrix.go:
```diff
@@ -7,36 +7,30 @@ import (
 		return nil, ok
 	}
-	return set.ToSlice(), ok
+	slice := make([]T, 0, set.Cardinality())
```
Because an `[]interface{}` value is not assignable to a variable of type `[]T`

akerouanton replied:
Ah, I see: we're still using an old version of `golang-set`. It supports generics since its v2.0 and `ToSlice()` now returns `[]T`. Maybe we should update it?

corhere replied:
Good catch! I completely overlooked that there was a v2 module

akerouanton reviewed (approved):
LGTM

cpuguy83 reviewed (approved):
Perfect use for generics

corhere committed:
libn/i/setmatrix: make generic and constructorless

Allow SetMatrix to be used as a value type with a ready-to-use zero
value. SetMatrix values are already non-copyable by virtue of having a
mutex field so there is no harm in allowing non-pointer values to be
used as local variables or struct fields. Any attempts to pass around
by-value copies, e.g. as function arguments, will be flagged by go vet.

Signed-off-by: Cory Snider <csnider@mirantis.com>
